### PR TITLE
Development

### PR DESCRIPTION
**Pull request by DerCalli**

### CODE REVIEW

File: src/FondOfSpryker/Yves/EnhancedEcommerce/EnhancedEcommerceDependencyProvider.php [hunk 1]
```diff
@@ -7,13 +7,86 @@
 
 class EnhancedEcommerceDependencyProvider extends AbstractBundleDependencyProvider
 {
+    public const ENHNACED_ECOMMERCE_DATALAYER_EXPANDER_PLUGINS = 'ENHNACED_ECOMMERCE_DATALAYER_EXPANDER_PLUGINS';
+    public const ENHNACED_ECOMMERCE_TWIG_PARAMETER_BAG_EXPANDER_PLUGINS = 'ENHNACED_ECOMMERCE_TWIG_PARAMETER_BAG_EXPANDER_PLUGINS';
+    public const ENHNACED_ECOMMERCE_CONTROLLER_EVENT_HANDLER = 'ENHNACED_ECOMMERCE_CONTROLLER_EVENT_HANDLER';
+
```
Comment (sourcecube) [inline src/FondOfSpryker/Yves/EnhancedEcommerce/EnhancedEcommerceDependencyProvider.php]: Namespacing container container constants.. `FOS_...`

File: src/FondOfSpryker/Yves/EnhancedEcommerce/Plugin/EnhancedEcommerceTwigServiceProvider.php [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+     * and should be used for "dynamic" configuration (whenever
+     * a service must be requested).
+     *
+     * @return void
+     */
+    public function boot(Application $app)
+    {
+        // TODO: Implement boot() method.
+    }
+}
```
Comment (sourcecube) [inline src/FondOfSpryker/Yves/EnhancedEcommerce/Plugin/EnhancedEcommerceTwigServiceProvider.php]: I believe this is deprecated and should be instead implementing `TwigPluginInterface`

File: src/FondOfSpryker/Yves/EnhancedEcommerce/Theme/default/partials/data-layer.twig [hunk 3]
```diff
@@ -0,0 +1,5 @@
+<script>
+    window.dataLayer = window.dataLayer || [];
+
+    {{ dataLayerString|raw }}
```
Comment (sourcecube) [inline src/FondOfSpryker/Yves/EnhancedEcommerce/Theme/default/partials/data-layer.twig]: Is this correct printing a plain string in script ? An object seems more logical, what's behind that dataLayerString variable?

Comment (DerCalli) [reply src/FondOfSpryker/Yves/EnhancedEcommerce/Theme/default/partials/data-layer.twig]: jepp, enhanced ecommerce is quite more then rendering a simple datalayer object. in some cases we need different renderings for the connector-modules (i.e. for javascript-stuff). before it was returning an array which was only rendering the datalayer. now the connector modules are returning complete rendered templates as string. the name "dataLayerString" is a little bit confusing, maybe javascript/javascriptString or something is a better choice.

File: src/FondOfSpryker/Yves/EnhancedEcommerce/EnhancedEcommerceDependencyProvider.php (hunk 1 above)
Comment (DerCalli) [reply src/FondOfSpryker/Yves/EnhancedEcommerce/EnhancedEcommerceDependencyProvider.php]: FOS?

Comment (sourcecube) [reply src/FondOfSpryker/Yves/EnhancedEcommerce/EnhancedEcommerceDependencyProvider.php]: fond of spryker

File: src/FondOfSpryker/Yves/EnhancedEcommerce/Theme/default/partials/data-layer.twig (hunk 3 above)
Comment (sourcecube) [reply src/FondOfSpryker/Yves/EnhancedEcommerce/Theme/default/partials/data-layer.twig]: Okay

File: src/FondOfSpryker/Yves/EnhancedEcommerce/Twig/EnhancedEcommerceTwigExtension.php [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+use Symfony\Component\HttpFoundation\Request;
+use Twig\Environment;
+use Twig_SimpleFunction;
+
+/**
+ * @method \FondOfSpryker\Yves\EnhancedEcommerce\EnhancedEcommerceFactory getFactory()
+ */
+class EnhancedEcommerceTwigExtension extends AbstractTwigExtensionPlugin
+{
+    public const FUNCTION_ENHANCED_ECOMMERCE = 'enhancedEcommerce';
```
Comment (julianhyatt) [inline src/FondOfSpryker/Yves/EnhancedEcommerce/Twig/EnhancedEcommerceTwigExtension.php]: maybe add 'enhancedEcommerce' to the shared constants?

File: src/FondOfSpryker/Yves/EnhancedEcommerce/Twig/EnhancedEcommerceTwigExtension.php [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+     * @param array $params
+     *
+     * @return string
+     */
+    public function renderEnhancedEcommerce(Environment $twig, string $page, Request $request, array $twigVariableBag): string
+    {
+        $renderedJavascript = '';
+        $dataLayer = [];
+
+        foreach ($this->getFactory()->getTwigParameterBagExpanderPlugins() as $twigVariableBagExpanderPlugin) {
```
Comment (julianhyatt) [inline src/FondOfSpryker/Yves/EnhancedEcommerce/Twig/EnhancedEcommerceTwigExtension.php:67]: Maybe extract every foreach to e separated function or plugin runner?

File: src/FondOfSpryker/Yves/EnhancedEcommerce/Plugin/Twig/EnhancedEcommerceTwigPlugin.php [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    /**
+     * @param Environment $twig
+     * @param ContainerInterface $container
+     *
+     * @return Environment
+     */
+    protected function registerEnhancedEcommerceTwigFunction(Environment $twig, ContainerInterface $container): Environment
+    {
+        $twig->addFunction(
+            new TwigFunction(static::TWIG_FUNCTION_ENHANCED_ECOMMERCE, null, ['needs_context' => true])
```
Comment (julianhyatt) [inline src/FondOfSpryker/Yves/EnhancedEcommerce/Plugin/Twig/EnhancedEcommerceTwigPlugin.php]: 'needs_context' as const

**Review by julianhyatt — COMMENTED** [review]
see comments

**Review by sourcecube — APPROVED** [review]

File: src/FondOfSpryker/Yves/EnhancedEcommerce/EnhancedEcommerceDependencyProvider.php [hunk 7]
```diff
@@ -7,13 +7,82 @@
...
+        $container = $this->addDataLayerExpanderPlugins($container);
+        $container = $this->addTwigParameterBagExpanderPlugins($container);
+        $container = $this->addRendererPlugins($container);
+
+        return $container;
+    }
+
+    protected function addRendererPlugins(Container $container): Container
+    {
+        $container->set(static::FOS_ENHNACED_ECOMMERCE_RENDERER_PLUGINS, function () {
```
Comment (daniel-rose) [inline src/FondOfSpryker/Yves/EnhancedEcommerce/EnhancedEcommerceDependencyProvider.php]: Use static function for closure (better performance)

File: src/FondOfSpryker/Yves/EnhancedEcommerce/EnhancedEcommerceDependencyProvider.php [hunk 8]
```diff
@@ -7,13 +7,82 @@
...
     {
-        return parent::provideDependencies($container); // TODO: Change the autogenerated stub
+        $container = $this->addDataLayerExpanderPlugins($container);
+        $container = $this->addTwigParameterBagExpanderPlugins($container);
+        $container = $this->addRendererPlugins($container);
+
+        return $container;
+    }
+
+    protected function addRendererPlugins(Container $container): Container
```
Comment (daniel-rose) [inline src/FondOfSpryker/Yves/EnhancedEcommerce/EnhancedEcommerceDependencyProvider.php:33]: Missing PHPDoc

File: src/FondOfSpryker/Yves/EnhancedEcommerce/EnhancedEcommerceDependencyProvider.php [hunk 9]
```diff
@@ -7,13 +7,82 @@
...
+    }
+
+    /**
+     * @param \Spryker\Yves\Kernel\Container $container
+     *
+     * @return \Spryker\Yves\Kernel\Container
+     */
+    protected function addDataLayerExpanderPlugins(Container $container): Container
+    {
+        $container->set(static::FOS_ENHNACED_ECOMMERCE_DATALAYER_EXPANDER_PLUGINS, function () {
```
Comment (daniel-rose) [inline src/FondOfSpryker/Yves/EnhancedEcommerce/EnhancedEcommerceDependencyProvider.php]: Use static function for closure (better performance)

File: src/FondOfSpryker/Yves/EnhancedEcommerce/EnhancedEcommerceDependencyProvider.php [hunk 10]
```diff
@@ -7,13 +7,82 @@
...
+    }
+
+    /**
+     * @param \Spryker\Yves\Kernel\Container $container
+     *
+     * @return \Spryker\Yves\Kernel\Container
+     */
+    protected function addTwigParameterBagExpanderPlugins(Container $container): Container
+    {
+        $container->set(static::FOS_ENHNACED_ECOMMERCE_TWIG_PARAMETER_BAG_EXPANDER_PLUGINS, function () {
```
Comment (daniel-rose) [inline src/FondOfSpryker/Yves/EnhancedEcommerce/EnhancedEcommerceDependencyProvider.php]: Use static function for closure (better performance)

**Review by daniel-rose — CHANGES_REQUESTED** [review]

**Review by daniel-rose — APPROVED** [review]

**Pull request merged**